tests/test-update-cleanup-workaround.c

```diff
@@ -0,0 +1,213 @@
+  ostree_ref_create (client_repo,
+                     ref,
+                     old_commit_id,
+                     &ref_created,
+                     &error);
+  g_assert_no_error (error);
+  g_assert_true (cmd_result_ensure_ok_verbose (&ref_created));
+}
+
+static void
+undeploy_old_deployment (EtcData *data)
+{
+  g_auto(CmdResult) undeployed = CMD_RESULT_CLEARED;
+  g_autoptr(GFile) client_sysroot = NULL;
+  g_autoptr(GError) error = NULL;
+
+  g_assert_nonnull (data);
+  g_assert_nonnull (data->client);
+
+  client_sysroot = eos_test_client_get_sysroot (data->client);
+  ostree_undeploy (client_sysroot,
+                   1,
+                   &undeployed,
+                   &error);
+
+  g_assert_no_error (error);
+  g_assert_true (cmd_result_ensure_ok_verbose (&undeployed));
+}
+
+static void
+delete_ref (EtcData *data,
+            const gchar *ref)
+{
+  g_autoptr(GFile) client_repo = NULL;
+  g_auto(CmdResult) ref_deleted = CMD_RESULT_CLEARED;
+  g_autoptr(GError) error = NULL;
+
+  g_assert_nonnull (data);
+  g_assert_nonnull (ref);
+  g_assert_nonnull (data->client);
+
+  client_repo = eos_test_client_get_repo (data->client);
+  ostree_ref_delete (client_repo, ref, &ref_deleted, &error);
+  g_assert_no_error (error);
+  g_assert_true (cmd_result_ensure_ok_verbose (&ref_deleted));
+}
+
+static void
+delete_some_old_dirtree_objects (EtcData *data)
+{
+  g_autoptr(GFile) client_repo = NULL;
+  OstreePruneFlags prune_flags = OSTREE_PRUNE_REFS_ONLY | OSTREE_PRUNE_NO_PRUNE | OSTREE_PRUNE_VERBOSE;
+  g_auto(CmdResult) listed = CMD_RESULT_CLEARED;
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GRegex) gregex = NULL;
+  g_autoptr(GMatchInfo) match_info = NULL;
+  guint removed;
+
+  g_assert_nonnull (data);
+  g_assert_nonnull (data->client);
+
+  client_repo = eos_test_client_get_repo (data->client);
+  ostree_prune (client_repo, prune_flags, 0, &listed, &error);
+  g_assert_no_error (error);
+  g_assert_true (cmd_result_ensure_ok_verbose (&listed));
+
+  gregex = g_regex_new ("Pruning\\s+unneeded\\s+object\\s+([0-9a-zA-Z]{64}.dirtree)",
+                        /* compile flags: */ 0,
+                        /* match flags: */ 0,
+                        &error);
+  g_assert_no_error (error);
+  g_regex_match (gregex,
+                 listed.standard_error,
+                 /* match flags: */ 0,
+                 &match_info);
+  for (removed = 0;
+       g_match_info_matches (match_info) && (removed < 3);
+       ++removed)
+    {
+      g_autofree gchar *object = g_match_info_fetch (match_info,
+                                                     /* match group: */ 1);
+      etc_delete_object (client_repo, object);
+      g_match_info_next (match_info, &error);
+      g_assert_no_error (error);
+    }
+  g_assert_cmpint(removed, >, 0);
+}
+
+static void
+test_update_cleanup_workaround (EosUpdaterFixture *fixture,
+                                gconstpointer user_data)
```

Comment:
Could do with a brief doc comment for the test function explaining what the test is trying to verify. Could also use `g_test_bug()`.